File: backend/schemas/clinical.py
"""Clinical domain schemas: orders, care events, vitals, monitoring, diagnostics."""
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, ConfigDict

from .hospital import AdmissionResponse, EncounterResponse


class ClinicalOrderCreate(BaseModel):
    encounter_id: Optional[int] = None
    patient_id: int
    doctor_id: Optional[int] = None
    department_id: Optional[int] = None
    order_type: str
    title: str
    priority: Optional[str] = "routine"
    notes: Optional[str] = None


class ClinicalOrderResponse(BaseModel):
    id: int
    facility_id: Optional[int] = None
    encounter_id: Optional[int] = None
    patient_id: int
    doctor_id: Optional[int] = None
    department_id: Optional[int] = None
    order_type: str
    title: str
    priority: str
    status: str
    notes: Optional[str] = None
    created_at: datetime
    completed_at: Optional[datetime] = None
    model_config = ConfigDict(from_attributes=True)


class CareEventResponse(BaseModel):
    id: int
    facility_id: Optional[int] = None
    patient_id: int
    actor_user_id: Optional[int] = None
    encounter_id: Optional[int] = None
    department_id: Optional[int] = None
    event_type: str
    title: str
    summary: Optional[str] = None
    severity: str
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class PatientTimelineResponse(BaseModel):
    encounters: List[EncounterResponse]
    admissions: List[AdmissionResponse]
    orders: List[ClinicalOrderResponse]
    events: List[CareEventResponse]


class VitalObservationCreate(BaseModel):
    patient_id: int
    encounter_id: Optional[int] = None
    department_id: Optional[int] = None
    source: Optional[str] = "manual"
    heart_rate: Optional[float] = None
    systolic_bp: Optional[float] = None
    diastolic_bp: Optional[float] = None
    spo2: Optional[float] = None
    temperature_c: Optional[float] = None
    respiratory_rate: Optional[float] = None
    observed_at: Optional[datetime] = None


class VitalObservationResponse(BaseModel):
    id: int
    facility_id: Optional[int] = None
    patient_id: int
    recorded_by_id: Optional[int] = None
    encounter_id: Optional[int] = None
    department_id: Optional[int] = None
    source: str
    heart_rate: Optional[float] = None
    systolic_bp: Optional[float] = None
    diastolic_bp: Optional[float] = None
    spo2: Optional[float] = None
    temperature_c: Optional[float] = None
    respiratory_rate: Optional[float] = None
    observed_at: datetime
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class MonitoringSignalResponse(BaseModel):
    id: int
    facility_id: Optional[int] = None
    patient_id: int
    vital_observation_id: Optional[int] = None
    encounter_id: Optional[int] = None
    department_id: Optional[int] = None
    signal_type: str
    severity: str
    title: str
    summary: str
    status: str
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class VitalSubmissionResponse(BaseModel):
    vital: VitalObservationResponse
    signals: List[MonitoringSignalResponse]


class DiagnosticResultCreate(BaseModel):
    order_id: int
    result_type: str
    title: str
    summary: str
    abnormal_flag: Optional[bool] = False
    status: Optional[str] = "final"


class DiagnosticReviewUpdate(BaseModel):
    review_status: Optional[str] = "reviewed"
    review_note: Optional[str] = None


class DiagnosticResultResponse(BaseModel):
    id: int
    facility_id: Optional[int] = None
    order_id: int
    encounter_id: Optional[int] = None
    patient_id: int
    doctor_id: Optional[int] = None
    department_id: Optional[int] = None
    result_type: str
    title: str
    summary: str
    abnormal_flag: bool
    status: str
    review_status: str
    review_note: Optional[str] = None
    reviewed_by_id: Optional[int] = None
    reviewed_at: Optional[datetime] = None
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)

File: rust_gateway/src/models/clinical.rs
use chrono::NaiveDateTime;
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

/// Clinical order model (lab, radiology, pharmacy, procedure, nursing).
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct ClinicalOrder {
    pub id: i64,
    pub facility_id: Option<i64>,
    pub encounter_id: Option<i64>,
    pub patient_id: i64,
    pub doctor_id: Option<i64>,
    pub department_id: Option<i64>,
    pub order_type: String,
    pub title: String,
    pub priority: String,
    pub status: String,
    pub notes: Option<String>,
    pub created_at: Option<NaiveDateTime>,
    pub completed_at: Option<NaiveDateTime>,
}

/// Care event model.
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct CareEvent {
    pub id: i64,
    pub facility_id: Option<i64>,
    pub patient_id: i64,
    pub actor_user_id: Option<i64>,
    pub encounter_id: Option<i64>,
    pub department_id: Option<i64>,
    pub event_type: String,
    pub title: String,
    pub summary: Option<String>,
    pub severity: String,
    pub created_at: Option<NaiveDateTime>,
}

/// Vital observation model with physiological indicators.
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct VitalObservation {
    pub id: i64,
    pub facility_id: Option<i64>,
    pub patient_id: i64,
    pub recorded_by_id: Option<i64>,
    pub encounter_id: Option<i64>,
    pub department_id: Option<i64>,
    pub source: String,
    pub heart_rate: Option<f64>,
    pub systolic_bp: Option<f64>,
    pub diastolic_bp: Option<f64>,
    pub spo2: Option<f64>,
    pub temperature_c: Option<f64>,
    pub respiratory_rate: Option<f64>,
    pub blood_glucose: Option<f64>,
    pub observed_at: Option<NaiveDateTime>,
    pub created_at: Option<NaiveDateTime>,
    pub is_deleted: i64,
    pub deleted_at: Option<NaiveDateTime>,
}

/// Real-time clinical monitoring signal.
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct MonitoringSignal {
    pub id: i64,
    pub facility_id: Option<i64>,
    pub patient_id: i64,
    pub vital_observation_id: Option<i64>,
    pub encounter_id: Option<i64>,
    pub department_id: Option<i64>,
    pub signal_type: String,
    pub severity: String,
    pub title: String,
    pub summary: String,
    pub status: String,
    pub created_at: Option<NaiveDateTime>,
}

/// Diagnostic lab/radiology result.
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct DiagnosticResult {
    pub id: i64,
    pub facility_id: Option<i64>,
    pub order_id: i64,
    pub encounter_id: Option<i64>,
    pub patient_id: i64,
    pub doctor_id: Option<i64>,
    pub department_id: Option<i64>,
    pub result_type: String,
    pub title: String,
    pub summary: String,
    pub abnormal_flag: i64,
    pub status: String,
    pub review_status: String,
    pub review_note: Option<String>,
    pub reviewed_by_id: Option<i64>,
    pub reviewed_at: Option<NaiveDateTime>,
    pub created_at: Option<NaiveDateTime>,
    pub is_deleted: i64,
    pub deleted_at: Option<NaiveDateTime>,
}

/// Spark streaming pipeline metrics.
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct SparkStreamingMetrics {
    pub id: i64,
    pub batch_id: i64,
    pub records_processed: i64,
    pub processing_time_ms: f64,
    pub ml_latency_ms: f64,
    pub timestamp: Option<NaiveDateTime>,
}
